src/Features/Core/Portable/EmbeddedLanguages/StackFrame/StackFrameLexer.cs

```diff
@@ -0,0 +1,368 @@
+        {
+            if (Position == Text.Length)
+            {
+                return default;
+            }
+
+            var startPosition = Position;
+            var ch = CurrentChar;
+            Position++;
+
+            while (IsNumber(ch))
+            {
+                ch = CurrentChar;
+                Position++;
+            }
+
+            var aritySpan = new TextSpan(startPosition, (Position - 1) - startPosition);
+            var arityToken = CreateToken(StackFrameKind.TextToken, Text.GetSubSequence(aritySpan));
+            return arityToken;
+        }
+
+        internal ImmutableArray<StackFrameToken> ScanArrayBrackets()
+        {
+            if (Position == Text.Length)
+            {
+                return default;
+            }
+
+            using var _ = ArrayBuilder<StackFrameToken>.GetInstance(out var builder);
+
+            while (Position < Text.Length)
+            {
+                // Use PreviousCharAsToken here because we expect the 
+                // start of the brackets to be the previous token that started
+                // the need to parse for an array
+                var token = PreviousCharAsToken();
+                if (!IsArrayBracket(token))
+                {
+                    break;
+                }
+
+                Position++;
+                builder.Add(token);
+            }
+
+            Debug.Assert(builder.Count >= 2);
+
+            return builder.ToImmutable();
+
+            static bool IsArrayBracket(StackFrameToken token)
+                => token.Kind is StackFrameKind.OpenBracketToken or StackFrameKind.CloseBracketToken;
+        }
+
+        public StackFrameToken PreviousCharAsToken()
+        {
+            var previousChar = Text[Position - 1];
+            return CreateToken(GetKind(previousChar), Text.GetSubSequence(new TextSpan(Position - 1, 1)));
+        }
+
+        internal StackFrameTrivia? ScanWhiteSpace(bool includePrevious)
+        {
+            if (Position == Text.Length)
+            {
+                return null;
+            }
+
+            var startPosition = includePrevious ? Position - 1 : Position;
+
+            while (IsBlank(CurrentChar))
+            {
+                Position++;
+            }
+
+            if (Position == startPosition)
+            {
+                return null;
+            }
+
+            var whitespaceSpan = new TextSpan(startPosition, Position - startPosition);
+            return CreateTrivia(StackFrameKind.WhitespaceTrivia, Text.GetSubSequence(whitespaceSpan));
+        }
+
+        public StackFrameToken CurrentCharAsToken()
+        {
+            if (Position == Text.Length)
+            {
+                return CreateToken(StackFrameKind.EndOfLine, VirtualCharSequence.Empty);
+            }
+
+            var previousChar = Text[Position];
+            return CreateToken(GetKind(previousChar), Text.GetSubSequence(new TextSpan(Position, 1)));
+        }
+
+        public bool IsAt(string val)
```

Comment:
this is unfortunately confusing given that `at` is something special in your language.   Consider LexerIsPositionedAtStartOf("...")